### DISCUSSION TITLE
Dto를 Record로 생성하는 것에 대해

### DISCUSSION THREAD

**Discussion by shkisme** [BDD-CLUB/01-doo-re-back#35] [논의 사항]
두레 백엔드 프로젝트에서, 모든 Dto를 Record로 생성하는 것에 대해 어떻게 생각하시나요?

제가 Dto에 Record 도입을 주장하는 이유는 다음과 같습니다.
- jdk 17 채택
- 불필요한 코드를 제거할 수 있다. (추가적으로 선언해주지 않아도) 
  - 멤버 변수는 private final로 선언된다.
  - 필드별 getter 자동 생성
  - 모든 멤버 변수를 인자로 하는 public 생성자를 자동으로 생성한다. (Dto Class에 매번 붙여주었던 `@Allargsconstructor` 제거 가능)

그러나 Record는 모든 필드에 대해서 public getter를 열어둔다는 점이 있고, 해당 getter의 형식이 `.getXXX()`가 아니라 `.XXX()`라서, Record로 선언한 객체를 다시 Class로 변경하면 해당 getter들을 모두 수정해줘야 한다는 점이 있습니다. (반대도 마찬가지) 그러나 이 점은 Record를 Dto에만 사용한다면 크게 문제가 되지 않는다고 생각합니다.

한편 @amaran-th 는 어떤 Dto는 Class고, 어떤 Dto는 Record면 코드를 읽을때도, 유지보수할 때도 혼란이 올 수 있을 것 같다는 의견을 주셨어요. 그래서 도입할거면, 컨벤션으로 정하는게 어떤가 합니다!

**Comment by amaran-th** [comment]
저도 이번에 record를 처음 사용해보았는데 굉장히 편리하더라구요.
보름님이 언급하신 이유로, dto로 record를 사용하는 것을 컨벤션으로 지정하는 것에 찬성합니다.

**Comment by 02ggang9** [comment]
저는 DTO에 Record 사용하는 것에 찬성합니다!

롬복 의존성을 제거하는 것이 마음에 들었습니다. 또, 어떤 책에서 읽은 기억이 나는데 DTO의 기능을 제한시킬 수 있다라는 점이 가장 큰 장점이라고 생각했습니다~!

+) class로 하면 데이터만 담는 dto에서 다른 행위를 할 수 있다고 생각해서 record로 강제성을 부여하는 점이 좋다고 생각했네요!

**Reply by shkisme** [reply]
오 그러네요! 좋은 의견 감사합니다. 다만 Record에서도 메서드 작성은 가능하여 추가적인 로직을 넣는 것까지 방지할 수는 없겠다는 생각이 듭니다. 하지만 불변을 보장해주기에 그 가치는 충분하다 싶네요! 👍👍

**Comment by JJimini** [comment]
저도 한 번 사용해보니 편하더라고요! 그런데 저도 `Record를 Dto에만 사용` 이 부분에만 동의합니다

**Comment by lcqff** [comment]
좋습니다😄

**Comment by 02ggang9** [comment]
만창일치로 record로 dto를 생성하는 것으로 결정되었습니다~!
